File: skeleton/src/main/java/me/shkschneider/skeleton/java/StringHelper.java
package me.shkschneider.skeleton.java;

import android.support.annotation.NonNull;
import android.telephony.PhoneNumberUtils;
import android.util.Patterns;

import java.text.Normalizer;
import java.util.ArrayList;
import java.util.List;
import java.util.Random;

import me.shkschneider.skeleton.helper.LogHelper;

public class StringHelper {

    public static final String ALPHA = "abcdefghijklmnopqrstuvwxyz";
    public static final String NUMERIC = "0123456789";
    public final static String HEX = NUMERIC + ALPHA.substring(0, 6);
    public static final String ALPHA_NUMERIC = ALPHA + NUMERIC;

    public static boolean nullOrEmpty(final String string) {
        return (string == null || (string.length() == 0));
    }

    public static String capitalize(@NonNull final String string) {
        return string.substring(0, 1).toUpperCase() + string.substring(1).toLowerCase();
    }

    public static String upper(@NonNull final String string) {
        return string.toUpperCase();
    }

    public static String lower(@NonNull final String string) {
        return string.toLowerCase();
    }

    public static boolean alpha(@NonNull final String string) {
        return chars(string.toLowerCase(), ALPHA);
    }

    public static boolean numeric(@NonNull final String string) {
        return chars(string.toLowerCase(), NUMERIC);
    }

    public static boolean alphanumeric(@NonNull final String string) {
        return chars(string.toLowerCase(), ALPHA_NUMERIC);
    }

    public static boolean url(@NonNull final String string) {
        return Patterns.WEB_URL.matcher(string).matches();
    }

    public static boolean email(@NonNull final String string) {
        return Patterns.EMAIL_ADDRESS.matcher(string).matches();
    }

    public static boolean phone(@NonNull final String string) {
        return PhoneNumberUtils.isGlobalPhoneNumber(string);
    }

    private static boolean chars(@NonNull final String string, @NonNull final String chars) {
        for (char c : string.toCharArray()) {
            if (! chars.contains(String.valueOf(c))) {
                return false;
            }
        }

        return true;
    }

    public static String[] split(@NonNull final String string) {
        return string.split("(?!^)");
    }

    public static String withoutAccents(@NonNull String string) {
        string = Normalizer.normalize(string, Normalizer.Form.NFD);
        // string = string.replaceAll("[^\\p{ASCII}]", ""); // ascii
        string = string.replaceAll("\\p{M}", ""); // unicode
        return string;
    }

    public static String[] split(@NonNull final String string, final char delimiter) {
        final List<String> list = new ArrayList<String>();
        final int size = string.length();
        int start = 0;
        for (int i = 0; i < size; i++) {
            if (string.charAt(i) == delimiter) {
                if (start < i) {
                    list.add(string.substring(start, i));
                }
                else {
                    list.add("");
                }
                start = i + 1;
            }
            else if (i == size - 1) {
                list.add(string.substring(start, size));
            }
        }
        final String[] array = new String[list.size()];
        list.toArray(array);
        return array;
    }

    public static String[] lines(@NonNull final String string, final boolean skipEmptyLines) {
        if (skipEmptyLines) {
            return string.split("[\n\r]+");
        }
        else {
            return string.split("\\r?\\n");
        }
    }

    public static String ellipsize(@NonNull final String string, final int maxLength) {
        if (string.length() > maxLength) {
            return string.substring(0, maxLength - 3) + "...";
        }
        return string;
    }

    public static String random(@NonNull final String characters, final int length) {
        if (length <= 0) {
            LogHelper.warning("Length was invalid");
            return null;
        }

        final Random random = new Random();
        final StringBuilder stringBuilder = new StringBuilder();
        for (int i = 0; i < length; i++) {
            stringBuilder.append(characters.charAt(random.nextInt(characters.length())));
        }
        return stringBuilder.toString();
    }

    public static String random(final int length) {
        return random(HEX.toLowerCase(), length);
    }

}

File: library/src/main/java/shkschneider/skeleton/java/StringHelper.kt
package me.shkschneider.skeleton.java

import android.support.annotation.IntRange
import android.util.Patterns
import me.shkschneider.skeleton.helper.LocaleHelper
import java.text.Normalizer
import java.text.NumberFormat

object StringHelper {

    // val EMPTY = ""
    val NULL = "null"
    val ALPHA = ('a' .. 'z').toString()
    val NUMERIC = ('0' .. '9').toString()
    val HEX = NUMERIC + ALPHA.substring(0, 6)

    fun hexadecimal(string: String): Boolean {
        return string.toLowerCase().all {
            it in ('a' .. 'f') || it.isDigit()
        }
    }

    fun url(string: String): Boolean {
        return string.matches(Patterns.WEB_URL.toRegex())
    }

    fun email(string: String): Boolean {
        return string.matches(Patterns.PHONE.toRegex())
    }

    fun phone(string: String): Boolean {
        return string.matches(Patterns.PHONE.toRegex())
    }

    fun count(string: String, s: String): Int {
        return string.count { s.contains(it) }
    }

    // <http://stackoverflow.com/a/3758880>
    fun humanReadableSize(@IntRange(from = 0) bytes: Long, binary: Boolean = true): String {
        val unit = if (binary) 1024 else 1000
        if (bytes < unit) {
            return bytes.toString() + " B"
        }
        val exp = (Math.log(bytes.toDouble()) / Math.log(unit.toDouble())).toInt()
        return String.format(LocaleHelper.Device.locale(),
                "%.1.toFloat() %sB",
                bytes / Math.pow(unit.toDouble(), exp.toDouble()), (if (binary) "KMGTPE" else "kMGTPE")[exp - 1] + if (binary) "i" else "")
    }

    // <https://stackoverflow.com/a/3322174/603270>
    fun withoutAccents(s: String): String {
        var string = s
        string = Normalizer.normalize(string, Normalizer.Form.NFD)
        // string = string.replaceAll("[^\\p{ASCII}]", ""); // ascii
        string = string.replace("\\p{M}".toRegex(), "") // unicode
        return string
    }

    fun number(number: Int): String {
        return NumberFormat.getNumberInstance(LocaleHelper.Device.locale()).format(number)
    }

    // <http://stackoverflow.com/a/9855338>
    fun hexadecimal(bytes: ByteArray): String {
        val result = StringBuffer()
        bytes.forEach {
            val octet = it.toInt()
            val firstIndex = (octet and 0xF0).ushr(4)
            val secondIndex = octet and 0x0F
            result.append(HEX[firstIndex])
            result.append(HEX[secondIndex])
        }
        return result.toString()
    }

}
